### PR TITLE
slack and snooze implementation

### PR DESCRIPTION
**Pull request by xavluiz**
full slack connect/disconnect and snooze/un-snooze demo

https://www.loom.com/share/d351f4ca44e249a18f80cf562e15b2e2

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -65,6 +66,18 @@
       {
         "command": "codetime.displayReadme",
         "title": "Code Time: Learn more"
+      },
+      {
+        "command": "codetime.connectSlack",
```
Comment (brettmstevens7) [inline package.json:71]: Could make these consistent, like `codetime.connectSlackWorkspace` and `codetime.disconnectSlackWorkspace`.

File: src/Util.ts [hunk 2]
```diff
@@ -165,420 +141,420 @@ export function findFirstActiveDirectoryOrWorkspaceDirectory(): string {
...
+  return plugin_uuid;
 }
 
 export function getAuthCallbackState() {
-    return fileIt.getJsonValue(getDeviceFile(), "auth_callback_state");
+  let auth_callback_state = fileIt.getJsonValue(getDeviceFile(), "auth_callback_state");
+  if (!auth_callback_state) {
+    auth_callback_state = uuidv4();
+    setAuthCallbackState(auth_callback_state);
+  }
```
Comment (brettmstevens7) [inline src/Util.ts:262]: When would we not get that value?

File: src/command-helper.ts [hunk 3]
```diff
@@ -1,357 +1,397 @@
...
+        shareSlackMessage(text);
+      } else {
+        window.showInformationMessage("Highlight and select text to share via Slack to continue.");
+      }
+    })
+  );
+
+  cmds.push(
+    commands.registerCommand("codetime.connectSlack", () => {
+      connectSlack();
```
Comment (brettmstevens7) [inline src/command-helper.ts:366]: Might do `connectSlackWorkspace` and `disconnectSlackWorkspace` here

File: src/managers/SlackManager.ts [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+  // lazily check if the user has completed the slack authentication
+  setTimeout(() => {
+    refetchSlackConnectStatusLazily(40);
+  }, 10000);
+}
+
+// disconnect slack flow
+export async function disconnectSlackAuth(authId) {
+  // get the domain
+  const integration = getSlackIntegrations().find((n) => n.authId === authId);
```
Comment (brettmstevens7) [inline src/managers/SlackManager.ts:69]: Might call this `getSlackWorkspaces`

File: src/managers/SlackManager.ts [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    return;
+  }
+  // ask before disconnecting
+  const selection = await window.showInformationMessage(
+    `Are you sure you would like to disconnect the '${integration.team_domain}' Slack workspace?`,
+    ...[YES_LABEL]
+  );
+
+  if (selection === YES_LABEL) {
+    await softwarePut(`/auth/slack/disconnect`, { authId }, getItem("jwt"));
```
Comment (brettmstevens7) [inline src/managers/SlackManager.ts:82]: What will happen if the request wasn't successful?

File: src/managers/SlackManager.ts [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+  const registered = await checkRegistration();
+  if (!registered) {
+    return;
+  }
+
+  const integrations = getSlackIntegrations();
+  let enabled = false;
+  for await (const integration of integrations) {
+    const web = new WebClient(integration.access_token);
+    const result = await web.dnd.setSnooze({ num_minutes: 30 }).catch((err) => {
```
Comment (brettmstevens7) [inline src/managers/SlackManager.ts:100]: We could make this 120 (2 hr) and a customizable VS Code setting

File: src/managers/SlackManager.ts [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+    dndInfo = await web.dnd.info().catch((e) => {
+      console.error("Error fetching slack do not disturb info: ", e.message);
+      return null;
+    });
+  }
+  return dndInfo;
+}
+
+// check if snooze is enabled for a slack workspace
+export async function isSlackSnoozeEnabled(domain) {
```
Comment (brettmstevens7) [inline src/managers/SlackManager.ts:197]: Might call this `isSlackDnDEnabled`

File: src/managers/SlackManager.ts [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+          integration["team_name"] = usersIdentify?.team?.name;
+          // add it
+          currentIntegrations.push(integration);
+
+          foundNewIntegration = true;
+        }
+      }
+    }
+
+    updateIntegrations(currentIntegrations);
```
Comment (brettmstevens7) [inline src/managers/SlackManager.ts:400]: Should this happen in a separate function to sync integrations?

**Review by brettmstevens7 — COMMENTED** [review]
Left a few comments

File: src/managers/SlackManager.ts [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+ * Post the message to the slack channel
+ * @param selectedChannel
+ * @param message
+ */
+function postMessage(selectedChannel: any, message: string) {
+  let slackAccessToken = getSlackAccessToken();
+  const web = new WebClient(slackAccessToken);
+  web.chat
+    .postMessage({
+      text: message,
```
Comment (pbstriker38) [inline src/managers/SlackManager.ts:419]: Is it possible to tell slack to post this a code snippet? Or do we just need to manually wrap the message in ```

Comment (xavluiz) [reply src/managers/SlackManager.ts:425]: ah, i forgot about that. i'll check

File: package.json (hunk 1 above)
Comment (xavluiz) [reply package.json:71]: updated

File: src/Util.ts (hunk 2 above)
Comment (xavluiz) [reply src/Util.ts:262]: > we null it out after we use it as it's specific for any auth flow

File: src/command-helper.ts (hunk 3 above)
Comment (xavluiz) [reply src/command-helper.ts:366]: updated

File: src/managers/SlackManager.ts (hunk 5 above)
Comment (xavluiz) [reply src/managers/SlackManager.ts:77]: yeah, that's a good question. I think we should remove it locally anyway

File: src/managers/SlackManager.ts (hunk 6 above)
Comment (xavluiz) [reply src/managers/SlackManager.ts]: updated

File: src/managers/SlackManager.ts (hunk 7 above)
Comment (xavluiz) [reply src/managers/SlackManager.ts]: updated

File: src/managers/SlackManager.ts (hunk 8 above)
Comment (xavluiz) [reply src/managers/SlackManager.ts]: not sure what you mean? i've updated the name but its meant to rewrite whatever is in that file

File: src/managers/SlackManager.ts (hunk 9 above)
Comment (xavluiz) [reply src/managers/SlackManager.ts:425]: i'll create a new PR for that one once i find out how

Comment (xavluiz) [reply src/managers/SlackManager.ts:424]: actually, its just markup like what you mentioned ``` to surround the text

**Pull request merged**